README.md
# ack-reload
[![build status](https://travis-ci.org/AckerApple/reload.svg)](http://travis-ci.org/AckerApple/reload)
[![NPM version](https://img.shields.io/npm/v/ack-reload.svg?style=flat-square)](https://www.npmjs.com/package/ack-reload)

Automatically refresh and reload your code in your browser when your code changes. No browser plugins required.

Why?
----

Restarting your HTTP server and refreshing your browser is annoying.

### Table of Contents
- [How does it work?](#how-does-it-work)
- [Benefits of ack-reload over original reload package](#benefits-of-ack-reload-over-original-reload-package)
- [Installation](#installation)
- [Examples](#examples)
  - [Stand-Alone Example](#stand-alone-example)
  - [Express Example](#express-example)
  - [Manually Reloading](#manually-reloading)
  - [Advanced Full Featured Example](#advanced-full-featured-example)
  - [Integrate with Existing App](#integrate-with-existing-app)
- [API](#api)
  - [Middlware](#middlware)
  - [Arguments](#arguments)
- [Commands](#commands)
- [Credits](#credits)
- [License](#license)

## How does it work?

ack-reload works in three different ways depending on if you're using it:

1. Run ack-reload in NodeJs process with full API control using config options
2. Use ack-reload middleware to "attach" ack-reload to a existing server route
3. As a command line tool which starts its own Server to monitor the file(s) you're editing for changes and to serve `reload-client.js` to the browser.

Once reload-server and reload-client are connected, the client side code opens a [WebSocket](https://en.wikipedia.org/wiki/WebSocket) to the server and waits for the WebSocket to close, once it closes, ack-reload waits for the server to come back up (waiting for a socket on open event), once the socket opens we reload the page.

## Benefits of ack-reload over original reload package
ack-reload is a fork of the npm package [reload](https://www.npmjs.com/package/reload)

- Html5Mode
  - ack-reload has the option to make all requests root bound
  - Original reload package does not play well with html5 routing apps
- onReload option
  - Everytime a websocket reload is issued a callback function can be called
- Port already in use prompt aka EADDRINUSE prompt
  - Added EADDRINUSE catcher that starts a cli-prompt when desired port is in use. Another port can be supplied to start server on another open port.
- Native Promises
  - When starting an ack-reload server via the API, a native promise is returned AFTER all services have been started
- Watching Files
  - Watching files is more intuitive and actually available outside of just the CLI
  - The original reload package does not auto watch files
- ack-reload does NOT have a name conflict on Ubuntu like reload does
- Better Logging
  - Better verbose logging where an outside library can mandate how logging occurs
- Better Client Script Inclusion
  - ack-reload package auto appends client script to all html requests

Express app for ack-reload
```javascript
var reload = require('ack-reload')
var app = express()
var publicDir = path.join(__dirname, 'public')

//Only host reload software. Create websocket to http attachment. File watching
app.use( reload.middleware(publicDir, server) )

http.createServer(app).listen(3000)
```

Express app for OLD reload
> The following code is how the OLD [reload](https://www.npmjs.com/package/reload) app worked
```javascript
var reload = require('reload')
var app = express()
var publicDir = path.join(__dirname, 'public')

app.get('/', function (req, res) {
  res.sendFile(path.join(publicDir, 'index.html'))
})

var server = http.createServer(app)

reload(server, app)

server.listen(3000)
```


## Installation

```bash
npm install [-g] [--save-dev] ack-reload
```

#### Typical Package Installation
```bash
npm install --save-dev ack-reload
```

#### Typical Global Installation
```bash
npm install -g ack-reload
```

## Examples

Three ways to use ack-reload
---

There are three different ways to use ack-reload.

1. As a server, allowing ack-reload to host your whole project
2. As middleware, allowing your exsting server project to utilize ack-reload
2. As a command line application to serve up static HTML files and be able to reload when the code is altered

Using ack-reload in Express
---
When used with Express ack-reload creates a new Express route for reload. When you restart the server, the client will detect the server being restarted and automatically refresh the page.

ack-reload can be used in conjunction with tools that allow for automatically restarting the server such as [supervisor](https://github.com/isaacs/node-supervisor) (recommended), [nodemon](https://github.com/remy/nodemon), [forever](https://github.com/nodejitsu/forever), etc.

### Stand-Alone Example
Serve and watch a folder all in one

**`server.js`:**
```javascript
var reload = require('ack-reload')
var publicDir = path.join(__dirname, 'public')

reload(publicDir,{port:3000})
```

### Express Example
Serve and watch a folder on an existing server

**`server.js`:**
```javascript
var express = require('express')
var http = require('http')
var path = require('path')
var reload = require('ack-reload')
var bodyParser = require('body-parser')
var logger = require('morgan')

var app = express()

var publicDir = path.join(__dirname, 'public')

app.set('port', process.env.PORT || 3000)
app.use(logger('dev'))
app.use(bodyParser.json()) //parses json, multi-part (file), url-encoded

var server = http.createServer(app)

// Reload code only on www url-path from public folder path
app.use("/www" reload.middleware(publicDir, server) )

// Another Reload code here only on admin url-path to admin folder path
var adminPath = path.join(__dirname,'admin')
app.use("/admin", reload.middleware(adminPath, server))

server.listen(app.get('port'), function(){
  console.log("Web server listening on port " + app.get('port'));
});
```

A demo template that has a change with every reload
**`public/index.html`:**
```html
<!doctype html>
<html>
  <head>
    <title>Reload Express Sample App</title>
  </head>
  <body>
    <h1>Reload Express Sample App</h1>
    <br />
    <div>Current Time: <script>document.write(Date.now())</script></div>
    <br />
    <div>Script Time: <script src="www/outputTime.js"></script></div>
  </body>
</html>
```

### Manually Reloading

Two ways to approach:
- Manually call returned reload function after starting a ack-reload server
- Integrate the reload function into running middleware

Manually Reloading Application
```javascript
var publicDir = path.join(__dirname, 'public')
var reload = require('ack-reload')

var options = {
  onReload:function(){
    console.log('reload occurred at '+new Date().toString())
  }
}

reload(publicDir, options)
.then(config=>{
  //force reload every 10 seconds of all browser websocket connections
  setInterval(config.reload, 10000);

  //stop reload services after 30 seconds
  setTimeout(function(){
    config.httpServer.close()
  }, 30000)
})
.catch(function(e){
  console.error(e)
})
```

Manually Reloading With MiddleWare

```javascript
var publicDir = path.join(__dirname, 'public')
var reload = require('ack-reload')

var server = http.createServer(function(){
  midware(req,res)
})

var midware = reload.middleware(publicDir,server)

//force reload every 10 seconds of all browser websocket connections
setInterval(midware.reload, 10000);

//stop reload services after 30 seconds
setTimeout(function(){
  server.close()
}, 30000)
```

### Advanced Full Featured Example

```javascript
require('ack-reload')(__dirname,{
  port:8080,
  log:console.log.bind(console),
  open:true,
  message:'reload',
  hostname:'127.0.0.1',
  filter:function(pathTo, stat){
    return stat.isDirectory() || pathTo.search(/\.(js|css|html)$/)>=0
  },
  startPage:'index2.html'
})
.then(function(config){
  //force reload every 10 seconds of all browser websocket connections
  setInterval(config.reload, 10000);

  //stop reload services after 30 seconds
  setTimeout(function(){
    config.httpServer.close()
  }, 30000)  
})
.catch(function(e){
  console.error(e)
})
```

### Integrate with Existing App
The following example takes and existing app, builds a new app, and server both together with ack-reload

Preexisting app.js
```javascript
const express = require('express');

const app = express();
// LOTS AND LOTS OF MIDDLEWARE HERE...
// server is not created since in this module.
app.use((err, req, res, next) => {
  // final error handler
});

module.exports = app;
```

serve.js
```javascript
const app = require('./app');
const http = require('http');
const reload = require('ack-reload');

//tell ack-reload folder to put watch on
const pathToWwwFiles = require('path').join(__dirname,'www');

//Our server that hosts app and reload code
const server = http.createServer(function(req,res){
  if( reload.isRequestForReload(req) ){
    midware(req, res)//html or reload.js request
  }else{
    app(req,res)
  }
});

//watch files, create websocket. Return function to process requests
const midware = reload.middleware(pathToWwwFiles, server)

// at this point we have server and app and reload code
server.listen(8888, () => { console.log('started!') });
```


## API

### Middlware

```javascript
var reload = require('ack-reload')
var midware = reload.middleware(pathTo)

require('http').createServer(function(req,res){
  midware(req,res)
})
.listen(8080,function(){
  if(err){
    console.log(err)
  }else{
    console.log('server started')
  }
})
```

### Arguments

- `pathTo`:  Folder to watch and serve. Defaults to current dir
- `options`: 
  - `port` Number = 8080 - The port to bind to. Can be set with PORT env variable as well.
  - `log` Function = console.log
  - `open` Boolean = true - open a browser window
  - `message` String - when port is in use, tailor prompt messages label
  - `hostname` String = localhost - This allows for custom hostnames. Defaults to localhost.
  - `filter` Function - function(pathTo,stat) when function returns true, file will be watched.
  - `startPage` String - Specify a start page. Defaults to index.html.
  - `log` Function = console.log - Method to process logging info.
  - `watch` Boolean = true - Enable/disable watching files. Manual reload will be required
  - `html5Mode` Boolean = false - Enable/disable always returning root index.html for all html 404 requests
  - `onReload` Function - Everytime a websocket reload is issued, have a function called

## Commands

Using ack-reload as a command line application
---

There are two ways to use the command line application.

1. In a directory serving blank static HTML files or
2. In a project with a `package.json` file

Each will require different modes of installing.

In case one you should install ack-reload globally with `npm install ack-reload -g`. Also with ack-reload installed globally you can go to any directory with an HTML file and use the command reload to constantly watch it and reload it while you make changes.

In case two you should install locally with `npm install --save-dev`, since this tool is to aid in development you should install it as a dev dependency.

Navigate to your html directory:

    ack-reload -b

This will open your `index.html` file in the browser. Any changes that you make will now reload in the browser. You don't need to modify your HTML at all.

### Usage for Command Line Application

```
Usage: ack-reload [options]

Options:

  -h, --help                        Output usage information
  -V, --version                     Output the version number
  -b, --browser                     Open in the browser automatically.
  -n, --hostname                    If -b flag is being used, this allows for custom hostnames. Defaults to localhost.
  -d, --dir [dir]                   The directory to serve up. Defaults to current dir.
  -e, --exts [extensions]           Extensions separated by commas or pipes. Defaults to html,js,css.
  -p, --port [port]                 The port to bind to. Can be set with PORT env variable as well. Defaults to 8080
  -s, --start-page [start-page]		  Specify a start page. Defaults to index.html.
  -v, --verbose						          Turns on logging on the server and client side. Defaults to true.
  - `html5Mode`                     Boolean = false - Enable/disable always returning root index.html for all html 404 requests
```

## Credits

This package was originally a fork and pull-request of [reload](https://www.npmjs.com/package/reload). My [pull request](https://github.com/jprichardson/reload/pull/62) turned out to be too large for the reload team to adopt. My pull request is now it's own package published on NPM.


## License

(MIT License)

Copyright 2016, JP Richardson  <jprichardson@gmail.com>


CHANGELOG.md
2.0.1 / 2017-08-30
------------------
Fixed html script inclusion when html code comments present

2.0.8 / 2017-07-12
------------------
Added node 8 tests
Fixed port 80 issue for websockets
Added onReload
Added timestamp during server reloads

2.0.7 / 2017-04-18
------------------
added .appveyor.yml test suite
enhanced reload.js script insertion

2.0.2 / 2017-04-18
------------------
added --html5Mode

2.0.1 / 2017-03-23
------------------
added process once SIGINT monitor killer

2.0.0 / 2017-02-17
------------------
Breaking Change: Include use of native Promise. NodeJs .10 is no longer supported
Removed a great amount of weight in dependencies. Package is far simpler to use and weighs far less
Removed Express as a dependency
Watching files is more intuitive and actually included outside of CLI
Better verbose logging where outside library can mandate how logging occurs
Added EADDRINUSE to cli-prompt where if port is in use, another port can be supplied to start server on another open port
Made more functional for all types of servers and not just an express server
reload package auto appends client script to all html requests

1.1.1 / 2017-01-28
------------------
Fixed undefined error log on send message. See: https://github.com/jprichardson/reload/pull/59

1.1.0 / 2016-11-12
------------------
Added client end web socket support for https. See: https://github.com/jprichardson/reload/pull/54

1.0.2 / 2016-10-31
------------------
Added error handling to websocket send. See: https://github.com/jprichardson/reload/pull/49

1.0.1 / 2016-07-15
------------------
Fixed onbeforeunload event not firing in reload-client See: https://github.com/jprichardson/reload/pull/46

1.0.0 / 2016-06-24
------------------
Added
- Two new badges to the README (code-style and npm version)
- Verbose mode as option for both Express and command line usage
- A sample app for express

Modified
- Re-wrote the README to reflect all of these changes
- Updated dependencies to their latest version’s
- Fixed race condition that caused reload to spam the server when using sockets for automatic reloading

Removed
- All delays (wait, normal, and socket) (Reload is now all automatic using web sockets (no delays at all))
- Client side sockjs web sockets (removed sockjs) (Now using native web sockets on the client side and ws on server side)

See: https://github.com/jprichardson/reload/pull/41

0.8.2 / 2016-06-24
------------------
- Fixed regression caused by Windows line endings. See: https://github.com/jprichardson/reload/pull/40

0.8.1 / 2016-06-05
------------------
- Allow reload from node server. See: https://github.com/jprichardson/reload/pull/38

0.8.0 / 2015-12-21
------------------
- fixed `hostname` flag. See: https://github.com/jprichardson/reload/pull/34
- use `exts` from command line. See: https://github.com/jprichardson/reload/pull/32

0.7.0 / 2015-10-21
------------------
- fixed `wait` flag: https://github.com/jprichardson/reload/pull/27

0.6.0 / 2015-10-12
------------------
- added `hostname/ -h` flag. See: https://github.com/jprichardson/reload/issues/14 and https://github.com/jprichardson/reload/pull/28

0.5.0 / 2015-09-28
------------------
- renamed `delay` flag to `reloadDelay`. See: https://github.com/jprichardson/reload/pull/26
- added `wait` flag. See: https://github.com/jprichardson/reload/pull/26

0.4.0 / 2015-08-17
------------------
- add `true` option to `delay` so that it waits indefinitely until server is up https://github.com/jprichardson/reload/pull/21
- express 4 routes, https://github.com/jprichardson/reload/pull/24

0.3.0 / 2015-07-17
------------------
- added option for start page. See: https://github.com/jprichardson/reload/pull/20

0.2.0 / 2015-06-29
------------------
* Added Express 4 Support

0.1.0 / 2013-09-30
------------------
* silence sockjs
* created `reload` bin that is useful for browser/html development

0.0.2 / 2013-03-14
------------------
* fixed bug that caused failure on hashbang urls
* set proper mime type on reload.js client side script

0.0.1 / 2013-03-13
------------------
* Initial release.


lib/reload.js
"use strict";

var path = require('path')
var fs = require('fs')
var WebSocket = require('ws')
var RELOAD_FILE = path.join(__dirname, './reload-client.js')
var reloadCode = fs.readFileSync(RELOAD_FILE, 'utf8')

var ackReloadLog = require('./log.function')
var http = require('http')
var nodeStatic = require('node-static')
var promisePrompt = require('./promisePrompt.function')
var open = require('open')
var watch = require('watch')
var filterMaker = require('./filterMaker.function')

module.exports = servePath
module.exports.reloadSocketByHttp = reloadSocketByHttp
module.exports.middleware = middleware

module.exports.isRequestForReload = isRequestForReload
//module.exports.isHtmlRequest = isHtmlRequest
//module.exports.isReloadJsRequest = isReloadJsRequest

/** Returns promise of Object{httpServer:Object, reload:Function, port:Number}
  @pathTo String = currentWorkingDirectory
  @options{
    port Number,
    log Function = console.log,
    open Boolean = true,
    message String - when port is in use, tailor prompt messages label
    filter Function - file watching filter,
    onReload Function - called with every reload
    watch Boolean = true - Enable/disable watching files. Manual reload will be required
  }
*/
function servePath(pathTo, options){
  pathTo = pathTo || process.cwd()//currentWorkingDirectory
  return createFolderWatchServer(pathTo, options)
}

/** returns Function(request, response) . Does not watch files, use reloadSocketByHttp to establish watch and websocket
  @pathTo String - path to serve
  @httpServer Object - Use to establish file watching and reload websocket
  @options Object - See servePath options
*/
function middleware(pathTo, httpServer, options){
  options = options || {}
  var fileRouter = new nodeStatic.Server(pathTo,{cache:false});

  var routes = function(req, res){
    if( isReloadJsRequest(req) ){
      return routes.sendScript(req, res, options)
    }
    
    if( isHtmlRequest(req, options.html5Mode) ){
      return routes.indexRequest(req,res)
    }
    
    fileRouter.serve(req,res)
  }
  
  routes.sendScript = function(req,res,options){
    res.setHeader('Content-Type','text/javascript')

    reloadCode = reloadCode.replace('socketUrl.replace()', 'socketUrl.replace(/(^http(s?):\\/\\/)(.*:)(.*)/,' + (options.port ? '\'ws$2://$3' + options.port : '\'ws$2://$3$4') + '\')')

    res.end(reloadCode)
  },
  
  routes.indexRequest = function(req,res){
    var reqFile = req.url.replace(/(.*\/)([^?]*)(\?.*)*/g,'$2')
    reqFile = reqFile || 'index.html'
    reqFile = reqFile.replace('/',path.sep)
    var readFile = path.join(pathTo,reqFile)

    respondHtmlFile(readFile, req, res)
    .catch(function(err){
      if(err.code=='ENOENT' && options.html5Mode){
        return respondHtmlFile(path.join(pathTo,'index.html'), req, res)
      }else{
        reqres404(req, res)
      }
    })
    .catch(function(e){
      ackReloadLog.error(e)
      reqres404(req, res)
    })
  }

  if(httpServer){
    routes.reload = reloadSocketByHttp(pathTo, httpServer, options)
  }

  return routes
}

function respondHtmlFile(readFile, req, res){
  return new Promise(function(res,rej){    
    fs.readFile(readFile,function(err,buff){
      if(err)return rej(err)
      res(buff)
    })
  })
  .then(function(buff){
    res.setHeader('Content-Type','text/html')
    res.setHeader('Cache-Control','no-cache, no-store, must-revalidate')
    res.setHeader('Pragma','no-cache')
    res.setHeader('Expires','0')

    var fileContents = injectScriptTag(buff.toString())

    res.end(fileContents)
  })
}

function injectScriptTag(rawString){
  var cutHints = rawString.match(/<!--(?!<!)[^\[>](.|\n|\r)*?-->/g)
  rawString = rawString.replace(/<!--(?!<!)[^\[>](.|\n|\r)*?-->/g,'<!--reload-replaced-hint-->')

  var insertIndex = rawString.search(/<script/)
  var tag = '<script src="reload/reload.js"></script>'


  if(insertIndex>0){
    rawString = rawString.substring(0,insertIndex) + tag + rawString.substring(insertIndex,rawString.length)
  }else{
    rawString = rawString + tag
  }
  
  //repair rawString hints that were stripped out
  while(cutHints && cutHints.length){
    rawString = rawString.replace('<!--reload-replaced-hint-->', cutHints[0])
    cutHints.shift()
  }

  return rawString
}

function reqres404(req, res){
  var reqFile = req.url.replace(/(.*\/)([^?]*)(\?.*)*/g,'$2')
  reqFile = reqFile || 'index.html'
  reqFile = reqFile.replace('/',path.sep)

  res.setHeader('Content-Type','text/plain')
  res.statusCode = 404
  res.status = 404
  
  return res.end('404 '+reqFile+' File Not Found')
}

/** return promise */
function createFolderWatchServer(outputFileFolder, options){
  options = options || {}
  options.port = options.port || 8080
  options.log = options.log || ackReloadLog
  options.open = options.open==null ? true : options.open

  var app = module.exports.middleware(outputFileFolder, null, options)

  return startServer(app, options.port, options)
  .then(function(config){
    if(config.httpServer){    
      config.reload = reloadSocketByHttp(outputFileFolder, config.httpServer, options)
      
      if(options.watch || options.watch==null){
        options.log('watching', outputFileFolder)
      }
  
      if(options.open){    
        options.hostname = options.hostname || 'localhost'
        var urlPath = 'http://'+options.hostname+':' + config.port + '/'
        if(options.startPage){
          urlPath += options.startPage
        }
        open(urlPath)
        options.log('serving '+urlPath)
      }
    }else{
      options.log('live reload server and file watching skipped')
    }

    return config
  })
}

function reloadSocketByHttp(pathTo, httpServer, options){
  options = options || {}

  var WebSocketServer = WebSocket.Server
  var wss = new WebSocketServer({ server: httpServer })

  var reloader = function(){
    if(options.onReload)options.onReload()

    wss.clients.forEach(function each(client){
      if (client.readyState === WebSocket.OPEN){
        client.send('reload')
      }
    })
  }

  if(options.watch || options.watch==null){
    watchPath(pathTo, reloader, options)
  }

  return reloader
}

/** returns watch.createMonitor. See npm watch */
function watchPath(pathTo, reloader, options){
  options = options || {}
  options.filter = options.filter || filterMaker('js','css','html')

  return watch.createMonitor(pathTo, options, function (monitor) {
    monitor.on("created", reloader)
    monitor.on("changed", reloader)
    monitor.on("removed", reloader)
    process.once('SIGINT', function(){monitor.stop()})
  })
}

/** returns promise Object{httpServer,port} */
function startServer(app, port, options){
  return new Promise(function(respond,reject){
    var httpServer = http.createServer(app)

    httpServer.once('error',function(e){
      reject(e)
    })

    httpServer.once('listening',function(e){
      respond({httpServer:httpServer, port:port})
    })

    httpServer.listen(port)
  })
  .catch(function(e){
    if(e.code=='EADDRINUSE'){
      return promptPortInUse(app, port, options)
    }

    throw e
  })
  .then(function(config){
    promisePrompt.stop()
    return config
  })
}

/** returns promise */
function promptPortInUse(app, port, options){
  return promisePrompt([{
    description:'Port '+port+' in use. Type new port number. 0 to stop:',
    name:'port',
    default:port+1,
    pattern: /^[0-9]+$/,
    message: 'Port must be a number',
  }],options)
  .then(function(res){
    var port = Number(res.port)
    if(port){
      return startServer(app, port, options)
    }
  })
}

function isHtmlRequest(req, html5Mode){
  var urlReview = req.url.replace(/\?.*/g,'')
  var index = urlReview ==='/' || urlReview.search(/[^?]*(\.html)/)>=0
  var noExt = html5Mode && urlReview.search(/^[^?]+(\.[^?]+)(\?|$)/)<0
  //var noExt = urlReview.search(/^[^?]+(\.[^?]+)(\?|$)/)<0
  return index || noExt ? true : false
}

function isReloadJsRequest(req){
  return req.url.toLowerCase().search(/reload\/reload\.js/)>=0
}

function isRequestForReload(req){
  return isHtmlRequest(req) || isReloadJsRequest(req)
}